src/compiler/scala/tools/nsc/Reporting.scala

```diff
@@ -30,17 +30,28 @@ trait Reporting extends scala.reflect.internal.Reporting { self: ast.Positions w
       def this(what: String, booleanSetting: Settings#BooleanSetting) {
         this(what, () => booleanSetting, booleanSetting)
       }
-      val warnings = mutable.LinkedHashMap[Position, String]()
-      def warn(pos: Position, msg: String) =
+      val warnings = mutable.LinkedHashMap[Position, (String, String)]()
+      def warn(pos: Position, msg: String, since: String = "") =
         if (doReport()) reporter.warning(pos, msg)
-        else if (!(warnings contains pos)) warnings += ((pos, msg))
+        else if (!(warnings contains pos)) warnings += ((pos, (msg, since)))
       def summarize() =
         if (warnings.nonEmpty && (setting.isDefault || doReport())) {
-          val numWarnings  = warnings.size
-          val warningVerb  = if (numWarnings == 1) "was" else "were"
-          val warningCount = countElementsAsString(numWarnings, s"$what warning")
-
-          reporter.warning(NoPosition, s"there $warningVerb $warningCount; re-run with ${setting.name} for details")
+          val warningsGroupedBySince = warnings.groupBy{case (pos, (msg, since)) => since}
+          val deprecationSummary = warningsGroupedBySince.size > 1
+          warningsGroupedBySince.map{warnings => 
+            val numWarnings   = warnings._2.size
+            val warningsSince = if (warnings._1.nonEmpty) s" (since ${warnings._1})" else ""
+            val warningVerb   = if (numWarnings == 1) "was" else "were"
+            val warningCount  = countElementsAsString(numWarnings, s"$what warning")
+            val rerun         = if (deprecationSummary) "" else s"; re-run with ${setting.name} for details"
+            reporter.warning(NoPosition, s"there $warningVerb $warningCount$warningsSince$rerun")
+          }
+          if (deprecationSummary) {
+            val numWarnings   = warnings.size
+            val warningVerb   = if (numWarnings == 1) "was" else "were"
+            val warningCount  = countElementsAsString(numWarnings, s"$what warning")
+            reporter.warning(NoPosition, s"there $warningVerb $warningCount in total; re-run with ${setting.name} for details")
+          }
```

Comment:
The idea is to give developers a better idea which deprecations will happen in the next version versus happening in the future. It also makes deprecations more useful for non-standard-library authors because they can prefix the version string with their library name. So it's not "20 deprecations", but "9 deprecations from the standard library which will be removed in 2.13, and 11 deprecations from library X targeted for version 5.0".